### DISCUSSION TITLE
Are the Magenta lib and Sound Banks loaded for each instance of the player?

### DISCUSSION THREAD

**Discussion by lautarodragan** [cifkao/html-midi-player#60] [General]
Hey there! First of all I want to say this is an awesome project. Thank you so much for building it and making it available for everyone.

Regarding my question: I've placed about 20 `<midi-player sound-font>`'s in my page, and the Memory footprint (as show by Google Chrome's Task Manager) grew to over 2.5gb! Event over 4.8gb at times (seems pretty random). And the page usually hangs indefinitely, too. I need to kill the tab from the Task Manager.

If I use no sound-font, it's only ~67mb. Without loading this script at all, it's ~46mb.

The RAM footprint seems to be related to the amount of instances. Some numbers 
- 7 instances: 974mb. 
- 5 instances: 568mb. 
- 3: 384mb. 
- 1: 230mb.

I could dynamically add and remove instances of the elements to limit the memory footprint, but UI and UX would suffer. Is there a way to globally load this library's dependencies, once for all instances?

**Comment by cifkao** [comment]
The Magenta library is loaded only once, but a separate `SoundFontPlayer` instance is created for each player and its [`loadSamples()`](https://github.com/magenta/magenta-js/blob/d8a76682abb0979b985e4b80f6b68b5123b9f8d5/music/src/core/player.ts#L494) method is called. This way, only the samples necessary for playing the sequence are loaded; however, this is done separately for each instance, which is probably why you are seeing this kind of memory footprint.

I think the best and probably also the easiest way to solve this would be to modify `SoundFontPlayer` (in Magenta) to use a shared instance of `SoundFont` instead of creating a new one for each player (provided that the SoundFont URL is the same). Otherwise a similar workaround could probably be implemented here.

**Reply by cifkao** [reply]
@lautarodragan For now you could perhaps try implementing a workaround on your end where you loop over your players (before actually loading content into them) and make sure `player.player.soundFont` is the same object for all of them, and let me know if that solves your problem.

**Reply by lautarodragan** [reply]
Thanks for answering @cifkao! I'll give that a try.
